---
title: 'Differential Gene Expression Analysis '
author: "Zilan Wen"
date: "2024-05-13"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1 Install R package, RStudio, GitHub

*  R and RStudio

Install R and install RStudio <https://posit.co/download/rstudio-desktop/> based on your operating system,  available for Linux, MacOS, and Windows 

Check the current R version. The current R version is R-4.4.0 

```{r,eval=FALSE} 
R.Version()#the code in the code chunk will not be run.
```


*  GitHub (optional)

GitHub is an online software development platform. It's used for storing, tracking, and collaborating on software projects. It makes it easy for developers to share code files and collaborate with fellow developers on open-source projects.

*  **Set up your GitHub**: Go here <https://github.com/>, log in or sign up with your email.

*  **Create your repository**: Repository is the project you can create which is on the top left of your GitHub home page.To create a new repository, you can click the green button `New`, then name the new repository.Choose Public or Private, and initialize this repository with add a README file. Next Click `Create repository`. The list of repositories is on your Github home page.

*  **Use the Github desktop version**: We are going to download and upload by commiting changes from our local files to the cloud. Go to GitHub Desktop<https://desktop.github.com/>, download based on your operating system. Then go to sign in with your GitHub account.Go back to your GitHUb repository, and click open with Github Desktop\
![click open with Github Desktop](C:/Users/timoi/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis/image/Github1.png){width=50%}

*  **Clone a repository and connect repository to local files**  In local path, choose where we want to save the repository in our local computer.click `Clone`. Now we have cloned the repository that we created on GitHub and we have connected it to the GitHub desktop application.\
![1](C:/Users/timoi/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis/image/Github2.png){width=50%}

* **upload local files to GitHub cloud** Any changes and updated files can be uploaded to GitHUb cloud though GitHub Desktop.Atfter change files, go back to GitHub Desktop and navigate the corresponding repository, you will have the changes listed on the top left. We now give the change a title such as '11' in the`summary (required)`box and click `Commit to main` on the bottom left, Then on the right, we can push comments to the origin remote by clicking `Push origin`.\
![local path](C:/Users/timoi/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis/image/Github3.png){width=50%}

*  **Clone a repository from other GitHUb accounts**
Forking a repository can make a copy of a repository from other account to your account.\

![local path](C:/Users/timoi/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis/image/Github5.png){width=50%}

# 2 Count normalization 
Created on 14.05.2024  
@author: Zilan Wen  
Email: zilan.wen@helsinki.fi 

## 2.1 Normalization Overview
The first step in the DE analysis workflow is count normalization, which is necessary to make accurate comparisons of gene expression between samples.

The main factors often considered during normalization are:

*  sequencing depth or coverage.

![local path](C:/Users/timoi/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis/image/Github4.png){width=50%}

*  Gene length

*  RNA composition

These three factors have been described very well in [DEG workshop Salmon] (https://github.com/hbctraining/DGE_workshop_salmon/blob/master/lessons/02_DGE_count_normalization.md)

# 2.1.1 Common normalization methods

*  `CPM` *counts per million*  counts scaled by total number of reads.

*  `TPM`*transcripts per kilobase million*  counts per length of transcript (kb) per million reads mapped.

*  `RPKM` or `RKPM` *reads/fragments per kilobase of exon per million reads / fragments mapped*  similar to TPM.
`DESeq2's median of ratios`  counts divided by sample-specific size factors determined by median ratio of gene counts relative to geometric mean per gene.


*  `EdgeR's trimmed mean of M values (TMM)` uses a weighted trimmed mean of the log expression ratios between samples.


The differences between RPKM, FPKM and TPM are explained clearly by [SatQuest video](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/)

# 2.1.2 DESeq2-normalized counts: Median of ratios method 

This method is explained clearly in [DEG_workshop_salmon](https://github.com/hbctraining/DGE_workshop_salmon/blob/master/lessons/02_DGE_count_normalization.md)
Since tools for differential expression analysis are comparing the counts between sample groups for the same gene, gene length does not need to be accounted for by the tool. However, `sequencing depth` and `RNA composition` need to be taken into account.

## 2.2 Count normalization of our dataset with DESeq2 
For transcriptome study, RNA samples which are isolated from control samples and treatment samples are sent to sequencing company. We will receive fastq files ater RNA sequencing. Then we check the quality of reads in fastq files, align reads to reference genome, and count reads per genes. Then we will get the raw reads count. In this experiment, we isolated RNA samples from Socts pines seedlings with three replicates (`Ha-1`,`Ha-2`, `Ha-3`) which were inoculated with Heterobasidion annosum, a root and stem rot disease pathogen. The seedling control is fungus-free (`Ctr-1`, `Ctr-2`, `Ctr-3`). Since genome of Scots pine is not available, we align reads to *Pinus taeda* genome which can be found in plantgenie <https://plantgenie.org/>.

In this tutorial we start with raw reads count ´.txt´ files, normalize the counts using DESeq2.This requires a few steps:

*  1. Install packages `DESeq2`, `EdgeR`

*  2. Build a file description `fileDesc.txt` file for all `.txt` files about counts matrix,including file path, inoculation, and description. 

*  3. Read data files.

*  4. Build `meta` data. `meta` file is similar to `filedesc`, bnut we use meta for create `DESeqDataset` object. Ensure the row names of the metadata dataframe are present and in the same order as the column names of the counts data frame.

*  5. Create `DESeqDataset` object

*  6. Generate size factors

*  7. Generate the normalized counts

### 2.2.1 Install packages

Install `DESeq2` package
```{r,eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
```

We need to install `EdgeR` package because we need the readDGE function here.
```{r,eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```
### 2.2.2 Build a file description 

Here the file description called `fileDesc` is already build and save in `data` folder. We will use it for reading a list of seperated files.

### 2.2.3 Read data files

Set to the current work directory
```{r}
setwd('C:/Users/timoi/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis')
setwd('C:/Users/Zilan/OneDrive - University of Helsinki/ZilanWen Github/RNA-seq_Differential-Gene-Expression-Analysis')
getwd()
```
Now we start to read raw data
```{r}
Coinfection.targets<-read.delim("./data/fileDesc.txt")
```
change the rawnames of the dataframe `Coinfection.targets`, later on we don't need to rename column names.
```{r}
rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
```
load the packege `edgeR`
```{r}
library(edgeR)
```
read the six `.txt` files listed in `Coinfection.targets`.
```{r}
Coinfection.orig <- readDGE(Coinfection.targets, header=F)
```
check the dimension of the data set
```{r}
dim(Coinfection.orig)
```
chreck the first 6 rows of the data

```{r}
head(Coinfection.orig)
```
Extact counts dataframe
```{r}
Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)
head(Coinfection.rawCount)
```
### 2.2.4 Build meta data
We define `sampletype`: we only have two types of samples, Control and Ha-infection
```{r}
sampletype <- factor(c(rep("Ha",3), rep("Ctr", 3)))
```
Build meta data frame
```{r}
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))
```
check the column name of counts dataframe
```{r}
colnames(Coinfection.orig$count)
```
check the rowname of meta dataframe
```{r}
rownames(meta)
```
Check that sample names match in both files
```{r}
all(colnames(Coinfection.orig$count) %in% rownames(meta))
```
### 2.2.5 Create `DESeqDataset` object

load the package `DESeq2`
Create DESEq2 dataset `dds`
```{r, message=FALSE}  
library(DESeq2)
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)
```
If you want to view dds, you can use the code:
```{r}
head(counts(dds))
```

### 2.2.6 Generate size factors
To perform the median of ratios method of normalization, DESeq2 has a single estimateSizeFactors() function that will generate size factors.
```{r}
dds <- estimateSizeFactors(dds)
```
take a look at the normalization factor applied to each sample 
```{r}
sizeFactors(dds)
```


### 2.2.7 Generate the normalized counts
Now, to retrieve the normalized counts matrix from dds, we use the counts() function and add the argument normalized=TRUE.
```{r}
normalized_counts <- counts(dds, normalized=TRUE)
```


Save the `normalized_counts`into results to your local path.
```{r,eval=FALSE}
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
```

# 3 Principal Component Analysis (PCA) 

